Use EDPutToken in ReducedEGProducer and modernize GEDPhotonCoreProducer

guitargeek opened the pull request:
#### PR description:

When working a bit with the ReducedEGProducer module in https://github.com/cms-sw/cmssw/pull/32462, I realized that this module produces many products and once. Using the `EDPutToken` instead of carrying the strings with product names around makes this a bit clearer I think, so this PR suggests to use the `EDPutToken` for all products.

Furthermore, the `GEDPhotonCoreProducer` module is modernized, and includes of `*Handle.h` are avoided in the package, because these includes are necessarily inherited from the inclusion of `Event.h` or `EventSetup.h`.

#### PR validation:

CMSSW compiles, local matrix tests pass.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport intended.

guitargeek committed:
use EDPutToken for output collections in ReducedEGProducer

guitargeek committed:
use EDPutToken for ValueMaps in ReducedEGProducer

guitargeek committed:
modernize GEDPhotonCoreProducer

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32626/20681

- This PR adds an extra 64KB to repository

cmsbuild commented:
A new Pull Request was created by  @guitargeek (Jonas Rembser) for master.

It involves the following packages:

RecoEgamma/EgammaPhotonProducers


@perrotta, @jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@Sam-Harper, @jainshilpi, @rovere, @lgray, @sobhatta, @lecriste, @afiqaize, @varuns23, @ram1123 this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9e8c87/12045/summary.html
**COMMIT**: bf295f4c16c3de37294db41dceebfa2f93058f3f
**CMSSW**: CMSSW_11_3_X_2021-01-09-1100/slc7_amd64_gcc900


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2716938
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

jpata commented on RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc:
```diff
@@ -326,152 +343,126 @@ ReducedEGProducer::ReducedEGProducer(const edm::ParameterSet& config)
-  auto hbheRecHits = std::make_unique<HBHERecHitCollection>();
-  auto photonPfCandMap = std::make_unique<edm::ValueMap<std::vector<reco::PFCandidateRef>>>();
-  auto gsfElectronPfCandMap = std::make_unique<edm::ValueMap<std::vector<reco::PFCandidateRef>>>();
+  reco::PhotonCollection photons{};
```
Could you help me understand if there's a specific reason to initialize the collections with {} (value-initialization with empty initializer list) rather than the default initializer? Just curious if it makes a difference for std::vector.

guitargeek replied:
To be honest I don't see the benefit of using the initializer list instead of the default constructor. I just used curly braces so often for constructor calls in this PR that it was just a habit. I removed the curly braces here now.

jpata commented on RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc:
```diff
@@ -326,152 +343,126 @@ ReducedEGProducer::ReducedEGProducer(const edm::ParameterSet& config)
-  }
-  for (const std::string& outid : outGsfElectronIds_) {
-    produces<edm::ValueMap<float>>(outid);
+  if (!ootPhotonT_.isUninitialized()) {
```
it seems to me that this may be merged with the if/else just above?

guitargeek replied:
Done.

jpata reviewed (commented):
Thanks for this refactoring. Perhaps it may be useful to run a memory profile just in case, to make sure there is no change just in case?

guitargeek committed:
implement comments and further refactoring

guitargeek commented:
Thanks for the review! I implemented the comments and made some little further refactors.

How should I do the memory profile? Also with igprof? I've never done that.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32626/20694

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #32626 was updated. @perrotta, @jpata, @cmsbuild, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9e8c87/12230/summary.html
**COMMIT**: cebc0bc37e19f0c143656a92e74da8b94f7b3fdb
**CMSSW**: CMSSW_11_3_X_2021-01-11-2300/slc7_amd64_gcc900


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2716944
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

guitargeek committed:
direct const access to ambiguousGsfTracks in GsfElectron

guitargeek commented:
In a discussion in another PR I learned that I should not remove the includes of `*Handle` if they are directly used:
https://github.com/cms-sw/cmssw/pull/32613#discussion_r555203779

To avoid further discussion here, I dropped the commit that did the same in this PR.

Instead, I added some more refactoring concerning the `ReducedEGProducer` which relates to the ambiguous GsfTracks in GsfElectrons.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32626/20707

- This PR adds an extra 52KB to repository

cmsbuild commented:
Pull request #32626 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9e8c87/12245/summary.html
**COMMIT**: 68ee21f475429b70ef5fcc20a0f5054971e3b2f5
**CMSSW**: CMSSW_11_3_X_2021-01-12-1100/slc7_amd64_gcc900


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2716932
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

jpata commented on RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc:
```diff
@@ -847,55 +822,48 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
+    // Also in this loop let's relink ambiguous tracks
     // And here we add them back, now from a new reduced collection:
-    for (const auto& it : ambigTracksInThisElectron) {
+    for (const auto& it : gsfElectron.ambiguousGsfTracks()) {
```
given the `clearAmbiguousGsfTracks` a few lines above, is the new version of the code indeed equivalent to the old?

guitargeek replied:
It seemed to me that the few deleted lines above just copied the `gsfElectron.ambiguousGsfTracks()` into a new vector `ambigTracksInThisElectron`, but we don't need to copy just for const access.

I agree, it seems suspicious because one wonders why the copying was done in the first place if it has no apparent purpose? I couldn't think of a reason.

guitargeek replied:
Absolutely not, you are right! I just see now the `gsfElectron.clearAmbiguousGsfTracks();` before. So the copying is needed. Will fix.

guitargeek replied:
I will fix the latest commit. Strange that it didn't show up in the matrix tests.

jpata reviewed (commented):
A new question arose when going over the updated code, please see inline.

To speed up convergence, I ran the memory comparison on the latest version myself, I notice no difference in CMSSW_11_3_X_2021-01-11-2300 using the wf 23434.21 using 50 events, so all looks good IMO:
- baseline: https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/32626/ig_mem_orig/16
- this PR: https://jpata.web.cern.ch/jpata/cgi-bin/igprof-navigator/32626/ig_mem_new/16

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32626/20720

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #32626 was updated. @perrotta, @jpata, @cmsbuild, @santocch, @slava77 can you please check and sign again.

jpata commented:
@cmsbuild please test

@guitargeek can you confirm that there's a way to pick up inadvertent changes to these modules with the tests? Do you have any private validation (possibly with more events) that could pick up the `ambiguousGsfTracks` stuff https://github.com/cms-sw/cmssw/pull/32626#discussion_r556558506, or other rare issues?

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9e8c87/12282/summary.html
**COMMIT**: c369b25a39367895d4e21eae41c8555d3a584ce0
**CMSSW**: CMSSW_11_3_X_2021-01-13-2300/slc7_amd64_gcc900


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2716967
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2716938
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 156 log files, 37 edm output root files, 37 DQM output files

jpata commented:
@guitargeek just to confirm, before signing off, I would like to make sure that you found out why the test didn't pick up the bug fixed in c369b25?
We may want to run a larger-stats validation just in case, to not miss rare issues.

guitargeek commented:
Hi @jpata, I agree that it's worth to check why the tests didn't pick up the bug I accidentally introduced. Maybe it even hints to a chunk of inactive code we can remove. This weekend I don't have time to look into this, but I'll come back to this PR next week.

guitargeek commented:
Hello @jpata, I validated now privately with workflow 10024.0 (2017 TTBar) and 1000 events. I looped over the electrons produced by the ReducedEGProducer (line 860 in the new version) and printed out kinematic info about the ambiguous GsfTracks, comparing the output with PR and without PR. The output was the same.

The thing is that not many events have GsfElectrons with ambiguous GSF tracks, so the bot probably didn't see any of these events. So the low stats were indeed the reason that the bug in the previous version of the commit was undetected.

jpata commented:
Thank you for the test! Will sign separately.

jpata commented:
+reconstruction
- technical, improves ReducedEGProducer and GEDPhotonCoreProducer
- tests show no difference, separate high-stats validation was reported in https://github.com/cms-sw/cmssw/pull/32626#issuecomment-762512087

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
